cmd/cloud-controller-manager/controller-manager.go

```diff
@@ -17,33 +17,282 @@ limitations under the License.
 // The external controller manager is responsible for running controller loops that
 // are cloud provider dependent. It uses the API to listen to new events on resources.
 
+// This file should be written by each cloud provider.
+// The current file demonstrate how other cloud provider should leverage CCM and it uses fake parameters. Please modify for your own use.
+
 package main
 
 import (
+	"errors"
+	"fmt"
 	"math/rand"
+	"net"
+	"net/http"
 	"os"
+	"strings"
 	"time"
 
+	"github.com/spf13/pflag"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/component-base/cli/flag"
 	"k8s.io/component-base/logs"
-	"k8s.io/kubernetes/cmd/cloud-controller-manager/app"
-
 	_ "k8s.io/component-base/metrics/prometheus/clientgo" // load all the prometheus client-go plugins
 	_ "k8s.io/component-base/metrics/prometheus/version"  // for version metric registration
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/cmd/cloud-controller-manager/app"
+	_ "k8s.io/kubernetes/cmd/cloud-controller-manager/app/config"
+	cloudcontrollerconfig "k8s.io/kubernetes/cmd/cloud-controller-manager/app/config"
+	"k8s.io/kubernetes/cmd/cloud-controller-manager/app/options"
+	cmconfig "k8s.io/kubernetes/cmd/controller-manager/app/config"
+	nodeipamcontroller "k8s.io/kubernetes/pkg/controller/nodeipam"
+	nodeipamconfig "k8s.io/kubernetes/pkg/controller/nodeipam/config"
+	"k8s.io/kubernetes/pkg/controller/nodeipam/ipam"
+	netutils "k8s.io/utils/net"
+)
+
+const (
+	// cloudProviderName shows an sample of using hard coded parameter
+	cloudProviderName = "SampleCloudProviderName"
+
+	// defaultNodeMaskCIDRIPv4 is default mask size for IPv4 node cidr
+	defaultNodeMaskCIDRIPv4 = 24
+	// defaultNodeMaskCIDRIPv6 is default mask size for IPv6 node cidr
+	defaultNodeMaskCIDRIPv6 = 64
 )
 
 func main() {
 	rand.Seed(time.Now().UnixNano())
 
-	command := app.NewCloudControllerManagerCommand()
+	// cloudProviderConfigFile shows an sample of parse config file from flag option
+	var flagset *pflag.FlagSet = pflag.NewFlagSet("flagSet", pflag.ContinueOnError)
+	var cloudProviderConfigFile *string = flagset.String("cloud-provider-configfile", "", "This is the sample input for cloud provider config file")
+	pflag.CommandLine.ParseErrorsWhitelist.UnknownFlags = true
+	_ = pflag.CommandLine.Parse(os.Args[1:])
+
+	cloud, err := cloudprovider.InitCloudProvider(cloudProviderName, *cloudProviderConfigFile)
+	if err != nil {
+		klog.Fatalf("Cloud provider could not be initialized: %v", err)
+	}
+	if cloud == nil {
+		klog.Fatalf("cloud provider is nil")
+	}
+
+	controllerList := []string{"cloud-node", "cloud-node-lifecycle", "service", "route"}
```

Comment:
note to self: this is an example of allow-listing specific controller loops

Reply:
I think a follow up PR which split out the base/vanilla version from the sample showing usage of the extensions would be a good idea. It makes it easier on cloud providers who can just copy the base version and reference the extension sample. 
If we put the samples in its own directory it would also cut down on the .import-restrictions allowing us to ensure we didn't allow any private dependencies in the base version.